### PR TITLE
Add parser for EditorConfig section names

### PR DESCRIPTION
**Pull request by agocke**
EditorConfig section names define a language for matching paths. This
commit introduces a parser for that language and a compiler to translate
that language to regular expressions.

Two globbing patterns supported by the language (`{...,...}` and `{num..num}`)
are not yet implemented. PROTOTYPE comment markers have been left there
to denote future work.

### CODE REVIEW

**Comment by etbyrd** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by agocke** [comment]
@dotnet/roslyn-compiler for review

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            // PROTOTYPE(editorconfig): Below remains unimplemented
+            //
+            // <choice> ::= "{" <choice-list> "}" 
+            // <choice-list> ::= <path-list> | <path-list> "," <choice-list>
+            // <range> ::= "{" <integer> ".." <integer> "}"
+            // <integer> ::= "-" <digit-list> | <digit-list>
+            // <digit-list> ::= <digit> | <digit> <digit-list>
+            // <digit> ::= 0-9
+
+            var sb = new StringBuilder();
```
Comment (333fred) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: >StringBuilder [](start = 25, length = 13)

`PooledStringBuilder.GetInstance()`?

Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Same deal as the other PR -- this code may be shared. I'm staying away from Roslyn-specific stuff for now.

**Comment by etbyrd** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by etbyrd** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Should we file a tracking bug or make a prototype comment to follow up if we determine that this won't be shared?

---
In reply to: [190944186](https://github.com/dotnet/roslyn/pull/27119#discussion_r190944186) [](ancestors = 190944186)

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                            return TokenKind.BadToken;
+                        }
+
+                        // Check for all of the possible escapes
+                        switch (_sectionName[nextPos])
+                        {
+                            case '\\':
+                                // "\\" -> "\"
+                                _position += 2;
+                                return TokenKind.Backslash;
```
Comment (333fred) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Not handled by the parser?

File: src/Compilers/Core/CodeAnalysisTest/EditorConfigTests.cs [hunk 3]
```diff
@@ -27,7 +29,7 @@ public void SimpleCase()
 [*.cs]
 my_prop = my_val
 
-", "Z:\\bogus");
```
Comment (333fred) [inline src/Compilers/Core/CodeAnalysisTest/EditorConfigTests.cs]: Why the change?

Comment (333fred) [reply src/Compilers/Core/CodeAnalysisTest/EditorConfigTests.cs]: Ah, now I see your comment below about on `/` being recognized as a dir separator.

---
In reply to: [191015436](https://github.com/dotnet/roslyn/pull/27119#discussion_r191015436) [](ancestors = 191015436)

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs (hunk 2 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Nice catch. And I clearly don't have a test, either. Will fix.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs (hunk 1 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Eh, since it's just perf I think we can address it as part of a larger perf investigation if it turns out to be important.

Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: i think we need at least a prototype comment to track this.

What is your non-roslyn envisioned usage? 

---
In reply to: [191026972](https://github.com/dotnet/roslyn/pull/27119#discussion_r191026972) [](ancestors = 191026972)

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(tokenKind);
+                }
+            }
+            sb.Append('$');
+
+            return sb.ToString();
+        }
+
+        private ref struct SectionNameLexer
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: >ef struct  [](start = 17, length = 10)

REF STRUCT FTW

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            sb.Append('^');
+            var lexer = new SectionNameLexer(sectionName);
+            while (!lexer.IsDone)
+            {
+                var tokenKind = lexer.Lex();
+                switch (tokenKind)
+                {
+                    case TokenKind.BadToken:
+                        // Parsing failure
+                        return null;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Consider using a Try pattern for this method if we have an explicit non-throwing failure case.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                    {
+                        // Backslash escapes the next character
+                        int nextPos = _position + 1;
+                        if (nextPos >= _sectionName.Length)
+                        {
+                            return TokenKind.BadToken;
+                        }
+
+                        // Check for all of the possible escapes
+                        switch (_sectionName[nextPos])
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: > switch (_sectionName[nextPos]) [](start = 23, length = 31)

Consider just a simple `if` here.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+                        // Matches just this character
+                        sb.Append(lexer.EatCurrentCharacter());
+                        break;
+                    case TokenKind.Question:
+                        // '?' matches any single character
+                        sb.Append('.');
+                        break;
+                    case TokenKind.Star:
+                        // Matches any string of characters except directory separator
+                        // Directory separator is defined in editorconfig spec as '/'
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: >// Directory separator is defined in editorconfig spec as '/' [](start = 24, length = 61)

I assume that windows developers will use \ irrespective of what the spec says. Where do we normalize the values or error out on bad slashes?

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+                        sb.Append("[^/]*");
+                        break;
+                    case TokenKind.StarStar:
+                        // Matches any string of characters
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Does that include a backslash?

**Comment by 333fred** [comment]
@agocke looks like there's a syntax error.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs (hunk 5 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: By "Try pattern" do you mean prefixing the method name with Try, or also using a bool return and an out var? Since `null` is never a valid value, seems like the bool/out pattern is overkill in this situation.

Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Try with null return si fine. 

---
In reply to: [191528129](https://github.com/dotnet/roslyn/pull/27119#discussion_r191528129) [](ancestors = 191528129)

**Comment by agocke** [comment]
@333fred Any further comments? I added https://github.com/dotnet/roslyn/projects/18#card-10158406 to track pooled object implementation.

**Comment by agocke** [comment]
test windows_determinism_prtest please

**Comment by etbyrd** [comment]
@dotnet test windows_debug_vs-integration_prtest please

**Comment by agocke** [comment]
@333fred ping
